src/generators.ts

```diff
@@ -43,3 +32,54 @@ export function last<T>(iter: Iterable<T>): T {
     }
     return val;
 }
+
+export class MemoizedIterable<T> {
+    private iterator: Iterator<T>;
+    private nextIndex: number;
+    private complete: boolean;
+
+    public constructor(iter: Iterator<T> | Iterable<T>) {
+        // case when iter is Iterable
+        if (iter[Symbol.iterator]) {
+            this.iterator = iter[Symbol.iterator]();
+        } else {
+            this.iterator = iter as Iterator<T>;
+        }
+        this.nextIndex = 0;
+        this.complete = false;
+    }
+
+    public get length(): number {
+        while (!this.complete) {
```

Comment:
If we did this in the constructor:

```js
this.consume = this.consumeIterator();
this.nextValue = this.consume.next;
```

You could do this loop like:

```js
for (const _ of {[Symbol.iterator]: this.consumer}) {}
```

I'm not sure about using that, thought I'd suggest it for discussion.